### PR TITLE
Check for NFD app at the startup

### PR DESCRIPTION
**Pull request by ashwinyogesh**
Check if NDN Forwarding Daemon is installed.
If installed -> Continue Regular Onboarding
If not installed -> Show a Toast Message and request User to install NDN

### CODE REVIEW

File: app/src/main/java/memphis/myapplication/IntroActivity.java [hunk 1]
```diff
@@ -1,25 +1,62 @@
...
-            Intent intent = new Intent(this, LoginActivity.class);
-            startActivity(intent);
+        Context context = getApplicationContext();
+        Session session = new Session(context);
+
+        // Check if NDN Forwarding Daemon is installed.
+        // If installed -> Continue Regular Onboarding
+        // If not installed -> Show a Message and request User to install NDN Forwarding Daemon.
+        PackageManager pm = context.getPackageManager();
+        if(isPackageInstalled(nfdAppPackageName,pm))
```
Comment (agawande) [inline app/src/main/java/memphis/myapplication/IntroActivity.java]: Spacing around if:
if (isPackageInstalled(nfdAppPackageName,pm)) {

File: app/src/main/java/memphis/myapplication/IntroActivity.java [hunk 2]
```diff
@@ -1,25 +1,62 @@
...
+            if (!session.getLoginStatus()) {
+                Intent intent = new Intent(this, LoginActivity.class);
+                startActivity(intent);
+            }
+            // we are logged in; take us to MainActivity
+            else {
+                Intent intent = new Intent(this, MainActivity.class);
+                startActivity(intent);
+            }
+        }else{
```
Comment (agawande) [inline app/src/main/java/memphis/myapplication/IntroActivity.java]: Spacing:
} else {

File: app/src/main/java/memphis/myapplication/IntroActivity.java [hunk 3]
```diff
@@ -1,25 +1,62 @@
...
+
+            Toast.makeText(context,"Please consider Installing NFD, this is " +
```
Comment (agawande) [inline app/src/main/java/memphis/myapplication/IntroActivity.java]: Maybe just "Please install NFD which is required for npChat to work. Start the NFD from within the NFD app and then reopen this app." Also space after comma: context, "..."

File: app/src/main/java/memphis/myapplication/IntroActivity.java [hunk 4]
```diff
@@ -1,25 +1,62 @@
...
 import android.content.Intent;
+import android.content.pm.PackageManager;
+import android.net.Uri;
 import android.os.Bundle;
 import android.support.v7.app.AppCompatActivity;
+import android.widget.Toast;
 
 public class IntroActivity extends AppCompatActivity {
 
+    final String nfdAppPackageName = "net.named_data.nfd";
```
Comment (agawande) [inline app/src/main/java/memphis/myapplication/IntroActivity.java]: Should this go in xml?

**Review by agawande — CHANGES_REQUESTED** [review]
We should also check if NFD is running? (Or actually that might be done by having a try-catch around face/face.processEvents - need to check what happens when application is started w/o NFD).

**Comment by ashwinyogesh** [comment]
> We should also check if NFD is running? (Or actually that might be done by having a try-catch around face/face.processEvents - need to check what happens when application is started w/o NFD).

One way to Solve This->
0. Have a method that returns defaultIdName in Globals.
`public static Name getDefaultIdName() { return defaultIdName; }`

1. Have a onResume method in MainActivity. onResume is used because, when application is re-opened from stack, we need to check for NFD status.
` @Override`
   ` protected void onResume() {`
        `super.onResume();`
       ` try {`
            `register_with_NFD(Globals.getDefaultIdName());`
       ` } catch (Exception e){`
            `e.printStackTrace();`
        `}`
    `}`

2. Modify register_with_NFD method to launch intent to NFD app when onRegisterFailure is triggered. Add the following in onRegisterFailure.
`Intent launchIntent = getPackageManager().getLaunchIntentForPackage(getString(R.string.nfd_package));`
                           ` if (launchIntent != null) {`
                                `Toast.makeText(getApplicationContext(), "Please Start NFD.",
                                        Toast.LENGTH_LONG).show();`
                                `startActivity(launchIntent); }`
                            `finish();`


If this one is an acceptable solution, I will add this code and submit a PR.

**Comment by agawande** [comment]
Hmm, solution seems okay - have you tested it?

**Comment by ashwinyogesh** [comment]
I have tested with Moto-g6Play and OnePlus 3. Also, it is better if we move request for media permissions to onResume method. If it is in onCreate, then the UI will be messed up for this scenario
1. User Installs the npChat app(given NFD is already installed but not running).
2. User goes to loginActivity. Clicks on register after filling details.
3. Since, NFD is not running, onRegistrationFailure is triggered and user is taken to NFD app and also at the same time the permissions are requested.
This kinda messes UI.

**Comment by agawande** [comment]
Okay, moving media permission should be fine.

**Comment by ashwinyogesh** [comment]
My recent commit has updates on everything we discussed. Should be able to merge :).

**Comment by jrandallclark** [comment]
@AshwinChandlapur Everything looks good. Could you please fix the conflicts caused by some of the other pull requests that have been merged?

**Comment by AshwinChandlapur** [comment]
Hey @jrandallclark, 
I tried fixing the conflicts and running the app. I ran into 
`Attempt to invoke virtual method 'net.named_data.jndn.util.MemoryContentCache memphis.myapplication.MemoryCache.getmCache()' on a null object reference`

In the method register_with_NFD, Globals.memoryCache.getmCache() is throwing null pointer. Also, in Globals setMemoryCache is unused.

**Comment by ashwinyogesh** [comment]
Yup @jrandallclark . I have resolved conflicts. I guess, the branch is ready to be merged.

**Comment by somenath1435** [comment]
@AshwinChandlapur squash commits

**Comment by jrandallclark** [comment]
@AshwinChandlapur It turns out that calling register_with_NFD within MainActivity's OnResume isn't a good idea as it causes the app to re-register with NFD every time the user sees the main activity. I've reverted your pr for now.

**Comment by AshwinChandlapur** [comment]
To fix this, we can have register_NFD method in onCreate and Permissions in onResume. I guess, that would fix this problem of re-registering.

**Comment by ashwinyogesh** [comment]
@jrandallclark 
The register_with_NFD method was put in onResume because of scenarios such as this
1. Log into the app with NFD status running.
2. While using the app, if user decides to clear the background apps then NFD app would be closed and app wouldn't inform user about the NFD.

By putting register_with_NFD in onResume method, we were checking if NFD is running everytime the user is in MainActivity.

We can discuss of other ways to implement, so we can avoid re-registering devices.

**Pull request merged**